packages/providers/src/utils/auto-retry-fetch.ts

```diff
@@ -0,0 +1,78 @@
+import type { ProviderOptions } from '../provider';
+
+import { sleep } from './sleep';
+
+type Backoff = 'linear' | 'exponential' | 'fixed';
+
+/**
+ * Retry options scheme
+ */
+export type RetryOptions = {
+  /**
+   * Amount of attempts to retry before failing the call.
+   */
+  maxRetries: number;
+  /**
+   * Backoff strategy to use when retrying. Default is exponential.
+   */
+  backoff?: Backoff;
+  /**
+   * Starting delay for backoff strategy. Default is 150ms.
+   */
+  baseDelay?: number;
+};
+
+/**
+ * Calculate the wait duration for the next retry attempt
+ * @param options - Retry options configuration
+ * @param retryAttemptNum - 1-based, tells which retry attempt is this
+ * @returns Next wait duration
+ */
+export function getWaitDelay(options: RetryOptions, retryAttemptNum: number) {
+  const duration = options.baseDelay ?? 150;
+
+  switch (options.backoff) {
+    case 'linear':
+      return duration * retryAttemptNum;
+    case 'fixed':
+      return duration;
+    case 'exponential':
+    default:
+      return 2 ** (retryAttemptNum - 1) * duration;
+  }
+}
+
+/**
+ * Returns a wrapped fetch function that will auto-execute itself in case of errors, until it succeeds
+ * @param fetchFn - Function to be auto-retried
+ * @param options - Retry options configuration
+ * @param retryAttemptNum - ZERO=first call, ONE=first retry, TWO=second retry, etc.
+ * @returns Whatever is the output of the `fetchFn` function
+ */
+export function autoRetryFetch(
+  fetchFn: NonNullable<ProviderOptions['fetch']>,
+  options?: RetryOptions,
+  retryAttemptNum: number = 0
+): NonNullable<ProviderOptions['fetch']> {
+  if (options === undefined) {
+    return fetchFn;
+  }
+
+  return async (...args) => {
+    try {
+      return await fetchFn(...args);
+    } catch (_error: unknown) {
```

Comment:
@arboleya I merged this but didn't realize that you didn't actually add the `ECONNREFUSED` handling back https://github.com/FuelLabs/fuels-ts/pull/1662#discussion_r1455721427. I think it's better to have it like it was previously until we figure out all the use cases where we want to retry instead of retrying on _every_ error. This implementation would unnecessarily repeat rejected bad requests and delay feedback to the user.

Reply:
Yes, maybe I wasn't clear enough, but I left it for you. 👍 

That's what I mean by:
 - ["[...] ensure you didn't forget to test anything after re-adding the `ECONNREFUSED` conditional"](https://github.com/FuelLabs/fuels-ts/pull/1662#discussion_r1456024812)